### PR TITLE
Support `map_filter` operator

### LINKED ISSUE #5273: [FEA]Support map_filter
I wish we can support map_filter function.

For example:
```
from pyspark.sql.types import *
from pyspark.sql.functions import *
jsonString="""{"Zipcode":704,"ZipCodeType":"STANDARD","City":"PARC PARQUE","State":"PR"}"""
df=spark.createDataFrame([(1, jsonString)],["id","value"])
df.show(truncate=False)
df2=df.withColumn("value",from_json(df.value,MapType(StringType(),StringType())))

spark.conf.set("spark.rapids.sql.regexp.enabled",True)
df3=df2.withColumn('mapfiltercol', expr(f""" map_filter(value, (k,v)  -> k rlike "Zip%"    )       """) )
df3.collect()
```

Not-supported-messages:
```
    ! <MapFilter> map_filter(from_json(MapType(StringType,StringType,true), value#12, Some(UTC)), lambdafunction(RLIKE(lambda k#70, Zip%), lambda k#70, lambda v#71, false)) cannot run on GPU because GPU does not currently support the operator class org.apache.spark.sql.catalyst.expressions.MapFilter

```

### PR DESCRIPTION
**Pull request by res-life**
Closes #5273

Support `map_filter` operator.

`map_filter` is used to filter entries in a map.

The usage of `map_filter` is: 
```
map_filter(map_column_expession, (key, value) -> prediate(key, value))
```
e.g.:
```
>>> df = spark.sql("SELECT map_from_arrays(array(1, 2, 3), array(1, 2, 3)) as map")
>>> df.show(truncate = False)
+------------------------+
|map                     |
+------------------------+
|{1 -> 1, 2 -> 2, 3 -> 3}|
+------------------------+

>>> df.selectExpr("map_filter(map, (key, value) -> key + value > 5) as ret").show(truncate = False)
+--------+
|ret     |
+--------+
|{3 -> 3}|
+--------+

```

This implementation is: 
1. apply the predicate function in `map_filter` to get a plain boolean column
2. convert the plain boolean column to a `list-of-boolean` column according to the offsets column of inputted map column
3. invoke cuDF `apply_boolean_mask` method to filter, Refer to https://github.com/rapidsai/cudf/pull/10773 

Signed-off-by: Chong Gao <res_life@163.com>

### CODE REVIEW

**Comment by res-life** [comment]
Depending on JNI PR https://github.com/rapidsai/cudf/pull/10812 to merge.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/higherOrderFunctions.scala [hunk 1]
```diff
@@ -530,4 +530,49 @@ case class GpuTransformValues(
...
+case class GpuMapFilter(argument: Expression,
+    function: Expression,
+    isBound: Boolean = false,
+    boundIntermediate: Seq[GpuExpression] = Seq.empty)
+    extends GpuMapSimpleHigherOrderFunction {
+
+  @transient lazy val MapType(keyType, valueType, valueContainsNull) = argument.dataType
+
+  // set parameter 'valueContainsNull' to the argument's `valueContainsNull`
+  override def dataType: DataType = MapType(keyType, valueType, valueContainsNull)
```
Comment (firestarman) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/higherOrderFunctions.scala:544]: NIT:
```suggestion
  override def dataType: DataType = argument.dataType
```

**Review by firestarman — COMMENTED** [review]
Looks good to me

**Review by revans2 — APPROVED** [review]
Agree with the one nit by @firestarman looks really good.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/higherOrderFunctions.scala (hunk 1 above)
Comment (res-life) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/higherOrderFunctions.scala]: Done.

**Comment by res-life** [comment]
build

**Pull request merged**